### DISCUSSION TITLE
How to run precached images

### DISCUSSION THREAD

**Discussion by og900aero** [gokcehan/lf#1634] [Q&A]
My problem is that when I want to look at new pictures, the drawing of the preview picture is a bit lengthy due to the code below:

```
image/*)
      orientation="$(identify -format '%[EXIF:Orientation]\n' -- "$file")"
      if [ -n "$orientation" ] && [ "$orientation" != 1 ]; then
        cache="$(hash "$file").jpg"
        cache "$cache" "$@"
        convert -- "$file" -auto-orient "$cache"
        draw "$cache" "$@"
      else
        draw "$file" "$@"
      fi
      ;;
```
What slows it down is part of the orientation instruction. I don't want to take it out, because it's a useful control function anyway.
Does anyone know of a method or script that, if I enter a path, caches the images for lf on the specified path?
I was thinking of something like, for example, using find to search for all the images, and for each result, it would somehow run the cache mechanism:
find /path/to/pictures -type f -name "*.jpg" -exec ???

**Comment by DusanLesan** [comment]
You mean something like this? https://github.com/LukeSmithxyz/voidrice/blob/master/.config/lf/scope

**Reply by og900aero** [reply]
This is the same as what is written for the lf previewer, only it uses a different program (inkscape). I don't need another previewer, but a shell script in which, if I enter a path, it recursively searches for all jpg files and caches them in lf. So somehow the necessary lines should be extracted from the lf previewer file and rewritten to some extent.

**Reply by DusanLesan** [reply]
Like this maybe: 
```
cmd on-cd &{{
	[ "$PWD" = "$XDG_CACHE_HOME/lf" ] && exit 0
	mkdir -p "$XDG_CACHE_HOME/lf"
	find . -maxdepth 1 -type f \( -iname \*.jpg -o -iname \*.jpeg -o -iname \*.png \) |
		parallel 'cache="$XDG_CACHE_HOME/lf/thumb.$(sha256sum <<< "{}" | cut -d " " -f 1).jpg"; [ -f "$cache" ] || convert -- "{}" -auto-orient "$cache"'
}}
```
`on-cd` is a hook that happens on every directory change(so it will not be applied for initial dir)
First we make sure we are not in cache dir (I think it could cause some infinite loops or something otherwise)
Then we make a directory that will be used for caching if needed
Find will go through some specific image formats in current dir
`parallel` will cache files in parallel (this example uses all threads but you can limit to only few like `-j 6`)

Of-course this is only an example so you can adjust it based on your needs

**Reply by og900aero** [reply]
Thx for your example, but unfortunately it doesn't work and rewriting it is beyond my abilities.

**Reply by og900aero** [reply]
It's my bad. The parallel command not default command in Debian 12, so I installed, and now working, if I run manually. Automaticcaly not run with on-cd. And one important problem: It does not turn the pictures that are turned in the right direction.

**Reply by DusanLesan** [reply]
You can always use simple for statement instead of parallel in case you want to avoid it. 
Can you check if on-cd is working for you?
Try
```
cmd on-cd &{{
	notify-send "on-cd" ">$PWD<"
}}
```

As for orientation issue, I cannot repro it since it did orient a few images when I tried this out yesterday. Running `convert` command (with the same parameters) on the file directly in terminal produces the correct image?

**Reply by og900aero** [reply]
Now the cached files appear in the cache folder if I start it manually, but the previewer also has to be rewritten, because it doesn't use the files in the cache, but draws it again by itself.
If I run the notify command under on-cd, it works fine also under lf, but find command with complete row...doesn't work.
lf -version show -> Go version: go1.19.8 :D So, I run now the following command: env CGO_ENABLED=0 go install -ldflags="-s -w" github.com/gokcehan/lf@latest
But the problem isnt gone.

**Reply by DusanLesan** [reply]
> Now the cached files appear in the cache folder if I start it manually, but the previewer also has to be rewritten, because it doesn't use the files in the cache, but draws it again by itself

Right. It needs to check if cache exists and show it, otherwise to show file directly

> If I run the notify command under on-cd, it works fine also under lf, but find command with complete row...doesn't work

So it seems that issue is not with `lf` calling `on-cd` but the command script itself. You can try picking it apart to find the cause.

- Like first cd into the directory with images and try: `find . -maxdepth 1 -type f \( -iname \*.jpg -o -iname \*.jpeg -o -iname \*.png \)`
- If `find` returns expected output, try using `convert` command in a same way it is used in script (of-course replace variables or make sure they are valid). 
- If that is ok, run something like`cache="$XDG_CACHE_HOME/lf/thumb.$(sha256sum <<< "{}" | cut -d " " -f 1).jpg"; [ -f "$cache" ] || convert -- "{}" -auto-orient "$cache"`, Where you would replace `{}` with a file name (and make sure XDG_CACHE_HOME is valid or replace it too)
- If that is ok too, try running the whole command (`find ... | parallel '... convert ...'`) from terminal

Tip: Try replacing parts of any lf script that does not work with `notify-send`. For example, before I even attempted to do real caching, I have tried
```
cmd on-cd &{{
	find . -maxdepth 1 -type f \( -iname \*.jpg -o -iname \*.jpeg -o -iname \*.png \) |
		while read -r image; do
			notify-send  "Image:" ">$image<"
		done
}}
```
and then started adding parts

**Reply by og900aero** [reply]
But how does the code in the previewer file know what name the code in the lfrc file created the given file cache? Is there any transition between the variables of the two files?

**Reply by DusanLesan** [reply]
The value of cache variable would be the same. Whatever you get as a cached name once when providing a real name, you should get any other time (if you have used the same name. Path included in one place but not in other would break it). So previewer would need to get the file name, and convert it to hashed name and that value would be the same as the one you have given to the cached file when creating it.

**Reply by DusanLesan** [reply]
I have tried out now to actually use the cache: 
`lfrc` file content
```
set previewer ~/.config/lf/preview
set cleaner ~/.config/lf/cleaner

cmd on-cd &{{
	[ "$PWD" = "$XDG_CACHE_HOME/lf" ] && exit 0
	mkdir -p "$XDG_CACHE_HOME/lf"
	find "$PWD" -maxdepth 1 -type f \( -iname \*.jpg -o -iname \*.jpeg -o -iname \*.png \) |
		parallel 'cache="$XDG_CACHE_HOME/lf/thumb.$(sha256sum <<< "{}" | cut -d " " -f 1).jpg"; [ -f "$cache" ] || convert -- "{}" -auto-orient "$cache"'
}}
```
`preview` file content:
```
#!/bin/sh

image() {
	if [ -n "$DISPLAY" ] && [ -z "$WAYLAND_DISPLAY" ]; then
		printf '{"action": "add", "identifier": "PREVIEW", "x": "%s", "y": "%s", "width": "%s", "height": "%s", "scaler": "contain", "path": "%s"}\n' "$4" "$5" "$(($2-1))" "$(($3-1))" "$1" > "$FIFO_UEBERZUG"
		exit 1
	else
		chafa "$1" -s "$4x"
	fi
}

setCache() {
	cache="$(readlink -f "$1" | sha256sum)"
	cache="$XDG_CACHE_HOME/lf/thumb.${cache%% *}.jpg"
}

case "$(file --dereference --brief --mime-type -- "$1")" in
	image/*)
		setCache "$1"
		if [ ! -f "${cache}" ]; then
			cache="$1"
			notify-send "Cache not found"
		fi
		image "${cache}" "$2" "$3" "$4" "$5" ;;
	*)
		bat -p --color=always "$1" --theme 'Visual Studio Dark+' --tabs 2 ;;
esac
exit 0
```

**Reply by og900aero** [reply]
Unfortunately, this doesn't work properly for me:

1. that line still doesn't work in on-cd
2. it doesn't clear the preview image if I go to an item that doesn't have a preview because it's not configured or a directory
3. it shows that it is not cached for all of them, even though I manually started caching on one of the folders

If I modify SetCache function to:

```
setCache() {
	cache="$(readlink -f "$1" | sha256sum)"
	cache="$XDG_CACHE_HOME/lf/thumb.${cache%% *}.jpg"
	notify-send "$1"
			notify-send "$cache"
}
```
Even then, I can see that the cache file generated by the previewer received a different value than the one generated by the find... line in lfrc.

**Reply by DusanLesan** [reply]
> that line still doesn't work in on-cd

Did you try taking commands apart? 
Try this change and go to directory with images:
```
cmd on-cd &{{
	[ "$PWD" = "$XDG_CACHE_HOME/lf" ] && exit 0
	mkdir -p "$XDG_CACHE_HOME/lf"
	find "$PWD" -maxdepth 1 -type f \( -iname \*.jpg -o -iname \*.jpeg -o -iname \*.png \) |
		parallel 'cache="$XDG_CACHE_HOME/lf/thumb.$(sha256sum <<< "{}" | cut -d " " -f 1).jpg"; notify-send "Cache:" ">$cache<"'
}}
``` 
and it should display a notification for each jpeg or png. If that is not ok, move to previous peace of command and see if it is executed. 

> it doesn't clear the preview image if I go to an item that doesn't have a preview because it's not configured or a directory

Not sure if your cleaner script is working, I did not notice any issues

 > it shows that it is not cached for all of them, even though I manually started caching on one of the folders

If `on-cd` (make sure it is called exactly like that) is not working for you, then the value you hash is not the same. Make sure it is the same with notify-send or logging it to a file.
```
cmd on-cd &{{
	[ "$PWD" = "$XDG_CACHE_HOME/lf" ] && exit 0
	mkdir -p "$XDG_CACHE_HOME/lf"
	find "$PWD" -maxdepth 1 -type f \( -iname \*.jpg -o -iname \*.jpeg -o -iname \*.png \) |
		parallel 'notify-send "File name to cache:" ">{}<"'
}}
```
```
setCache() {
	notify-send "$1"
	notify-send "$(readlink -f "$1")"
}
```
Compare what you get for some file an make sure they match exactly

Did you try minimal setup like mine? So nothing else in these files. Same content like I had.

**Reply by og900aero** [reply]
I try minimal setup. The only settings have in lfrc:

```
set shell sh
set previewer ~/.config/lf/previewer
set cleaner ~/.config/lf/cleaner
set shellopts '-eu'
set ifs "\n"

cmd on-cd &{{
	[ "$PWD" = "$XDG_CACHE_HOME/lf" ] && exit 0
	mkdir -p "$XDG_CACHE_HOME/lf"
	find "$PWD" -maxdepth 1 -type f \( -iname \*.jpg -o -iname \*.jpeg -o -iname \*.png \) |
		parallel 'cache="$XDG_CACHE_HOME/lf/thumb.$(sha256sum <<< "{}" | cut -d " " -f 1).jpg"; [ -f "$cache" ] || convert -- "{}" -auto-orient "$cache"'
}}
```

Previewer file:

```
#!/bin/sh

image() {
	if [ -n "$DISPLAY" ] && [ -z "$WAYLAND_DISPLAY" ]; then
		printf '{"action": "add", "identifier": "PREVIEW", "x": "%s", "y": "%s", "width": "%s", "height": "%s", "scaler": "contain", "path": "%s"}\n' "$4" "$5" "$(($2-1))" "$(($3-1))" "$1" > "$FIFO_UEBERZUG"
		exit 1
	else
		chafa "$1" -s "$4x"
	fi
}

setCache() {
	cache="$(readlink -f "$1" | sha256sum)"
	cache="$XDG_CACHE_HOME/lf/thumb.${cache%% *}.jpg"
	notify-send "$1"
			notify-send "$cache"
}

case "$(file --dereference --brief --mime-type -- "$1")" in
	image/*)
		setCache "$1"
		if [ ! -f "${cache}" ]; then
			cache="$1"
			notify-send "Cache not found"
		fi
		image "${cache}" "$2" "$3" "$4" "$5" ;;
	*)
		bat -p --color=always "$1" --theme 'Visual Studio Dark+' --tabs 2 ;;
esac
exit 0
```

But the hash generated by lfrc (manually run in example directory) and previewer is still different.

When on-cd command contains only the following, then run properly

```
cmd on-cd &{{
	[ "$PWD" = "$XDG_CACHE_HOME/lf" ] && exit 0
	mkdir -p "$XDG_CACHE_HOME/lf"
	find "$PWD" -maxdepth 1 -type f \( -iname \*.jpg -o -iname \*.jpeg -o -iname \*.png \) |
		parallel 'notify-send "File name to cache:" ">{}<"'
}}
```

**Reply by DusanLesan** [reply]
Is the notification from `on-cd` the same for some file as when you preview it and print value to be hashed? For hash to be different, the value that is hashed must be different. Either path is included in one and not other, or extension is not added properly, or your shell handles preview code differently (`#!/bin/sh` for me is `zsh`). Make sure they are the same before caching in both places.

**Reply by og900aero** [reply]
Thanks for the help, but I'd rather give up on this. It's not worth as much as it is worth.
Thanks again for everything.
